src/utils.ts

```diff
@@ -0,0 +1,78 @@
+'use strict'
+
+import { Snippet } from './extension'
+
+interface SnippetGrammar {
+  tokens: Token[]
+}
+
+interface Token {
+  snippet: Pattern
+  variable: Pattern
+}
+
+interface Pattern {
+  re: RegExp
+  replacement: string
+}
+
+export const TABSTOP: Token = {
+  snippet: {
+    re: /\$(\d)/g,
+    replacement: 'TABSTOP_$1',
+  },
+  variable: {
+    re: /TABSTOP_(\d)/g,
+    replacement: '$$$1',
+  },
+}
+
+export const PLACEHOLDER: Token = {
+  snippet: {
+    re: /\$\{(\d):(\w+)\}/g, // tabstop : default value
+    replacement: 'PLACEHOLDER_$1_$2',
+  },
+  variable: {
+    re: /PLACEHOLDER_(\d)_(\w+)/g,
+    replacement: '${$1:$2}',
+  },
+}
+
+export function parseSnippets(
+  snippets: { [name: string]: Snippet },
```

Comment:
Should `snippets` have defined their interface?

Reply:
This seems good to me.

Reply:
Great. Would not defining it as return type become cumbersome, however?

Reply:
That's also true.

Reply:
The interface is defined now.